### DISCUSSION TITLE
Separating `Block` from `Proposal`

### DISCUSSION THREAD

**Discussion by colibrishin** [planetarium/libplanet#2838] [General]
We have a marshaled `Block` in the `Proposal`. The previous decision was made due to a tricky proposed block-transferring timing situation. This should be separated at some point in my opinion and should be considered with specification changes.

## Block-transferring
A node needs to receive a block and the corresponding proposal before the Propose timeout to validate a proposed block and vote on it. The separation causes the change in the behavior of voting at the Propose -> PreVote stage.

https://github.com/planetarium/libplanet/blob/7a7faba7354f74728fa8a54562e1b8035c66d2bd/Libplanet.Net/Consensus/Context.cs#L413-L424
https://github.com/planetarium/libplanet/blob/7a7faba7354f74728fa8a54562e1b8035c66d2bd/Libplanet.Net/Consensus/Context.Mutate.cs#L36-L44
https://github.com/planetarium/libplanet/blob/7a7faba7354f74728fa8a54562e1b8035c66d2bd/Libplanet.Net/Consensus/Context.Mutate.cs#L125-L146
https://github.com/planetarium/libplanet/blob/7a7faba7354f74728fa8a54562e1b8035c66d2bd/Libplanet.Net/Consensus/MessageLog.cs#L216-L224

## Evidence
Evidence is data for convicting the validator whose acts possibly lead to the fault of the consensus, and the example of those are the double vote (exceeding total quorum) and double proposal (minority proposal received validator will fall behind from the consensus).

Assuming that 
- Any double vote or proposal created from a node is abnormal.
- If the signatures of the vote or proposal are valid.

I think this could be used as evidence and punish the validator later on. _(Speculation) We can simply put the vote or proposal into the evidence list and validates the signature and compare it with the local evidence._

However, If we add the evidence into the `Block` and `Proposal` will have the marshaled block and that will cause the enlarging of the block size.

**Comment by colibrishin** [comment]
We could change the behavior of `GetProposal()` in `Context<T>` to return only if the block exists in the `MessageLog`. If a node didn't receive a block and proposal, then the node cannot PreVote-ing if the block is valid or not, but timeout NIL can be done. I think it will be the least fixing time or effort method that I have in mind.

This method has some possible problems we might encounter, for example, Let's assume we would accept the block message when the proposal is in the `MessageLog`. If the block message has arrived before the proposal, then it will be hidden for a brief moment by gossip caching.

**Comment by riemannulus** [comment]
I have some questions for before about this suggestion.
1. Why `Evidence` is added in `Block<T>`? Could we use another store?
2. Could you suggest the struct of `Evidence` and how to handle it? (e.g., https://docs.cosmos.network/v0.45/modules/evidence/01_concepts.html)
3. Is this suggestion critical enough to be discussed during a PoS transition? Is it impossible to separate later?

**Comment by riemannulus** [comment]
Also, I have some questions about this suggestion.
1. So, what do you think about how to separate `Block` from `Proposal`? As you said, we have a block sync timing issue if just separate from `Proposal`. I wonder your opinion on how to solve this problem.
2. Is storing all `ConsensusMessage` the only way to punish malicious nodes?

**Comment by ilgyu-yi** [comment]
I think this idea is good, though this can be huge chunk of work.
This can be done by separating uncommitted block sync logic from proposal,
but let this issue as future work would be better for now, since its advantage seems to be relatively small to effect of other tasks.
I agree to do this someday.
